File: Leetcode Problems Solutions/Java/46. Permutations.java
/*
Problem Link: https://leetcode.com/problems/permutations/
*/

import java.util.*;

class Solution {

    private void helper(List<List<Integer>> result, List<Integer> temp, int[] nums) {
        if (temp.size() == nums.length) {
            result.add(new ArrayList<>(temp));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (temp.contains(nums[i])) {
                continue;
            }
            temp.add(nums[i]);
            helper(result, temp, nums);
            temp.remove(temp.size() - 1);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        helper(result, temp, nums);
        return result;       
    }
}

File: Problems/46. Permutations.py
"""
Problem Link: https://leetcode.com/problems/permutations/
"""


class Solution:

    def permute_helper(self, nums, curr, result):
        if (len(nums) == 0):
            result.append(curr)
            return
        for i in range(len(nums)):
            current_digit = nums[i]
            remaining_digits = nums[:i] + nums[i+1:]
            self.permute_helper(remaining_digits, curr +
                                [current_digit], result)

    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.permute_helper(nums, [], result)
        return result
